en, zh: add upgrade TiDB Operator offline (#1200)
pingcap/docs-tidb-operator: en/upgrade-tidb-operator.md

This document describes how to upgrade TiDB Operator and Kubernetes.

> **Note:**
>
> You can check the currently supported versions of TiDB Operator using the `helm search repo -l tidb-operator` command.
> If the command output does not include the latest version, update the repo using the `helm repo update` command. For details, refer to [Configure the Help repo](tidb-toolkit.md#configure-the-helm-repo).

## Upgrade TiDB Operator online

1. Update CustomResourceDefinition (CRD) for Kubernetes. For more information about CRD, see [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/).

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/v1.2.0-beta.2/manifests/crd.yaml && \
    kubectl get crd tidbclusters.pingcap.com
    ```

> 2. Get the `values.yaml` file of the `tidb-operator` chart for the new version of TiDB Operator.

Reviewer: ```suggestion
2. Get the `values.yaml` file of the `tidb-operator` chart for the new TiDB Operator version. 
```

> 2. Get the `values.yaml` file of the `tidb-operator` chart for the new version of TiDB Operator.

Reviewer: ```suggestion
2. Get the `values.yaml` file of the `tidb-operator` chart for the TiDB Operator version.
```

    kubectl get po -n tidb-admin -l app.kubernetes.io/instance=tidb-operator -o yaml | grep 'image:.*operator:'
    ```

    If TiDB Operator is successfully upgraded, the expected output is as follows. `v1.2.0-beta.2` represents the desired version of TiDB Operator.

    ```
    image: pingcap/tidb-operator:v1.2.0-beta.2
    image: docker.io/pingcap/tidb-operator:v1.2.0-beta.2
    image: pingcap/tidb-operator:v1.2.0-beta.2
    image: docker.io/pingcap/tidb-operator:v1.2.0-beta.2
    ```

    > **Note:**
    >

>     > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will be automatically upgraded to the specified version of TiDB Operator.

Reviewer: ```suggestion
    > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will be automatically upgraded to the corresponding version of TiDB Operator.
```

## Upgrade TiDB Operator offline

If your server cannot access the Internet, you can take the following steps to upgrade TiDB Operator offline:

1. Download the files and images required for the upgrade using a machine with the Internet access:

    1. Download the `crd.yaml` file for the new TiDB Operator version. For more information about CRD, see [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom- resource-definitions/).

    {{< copyable "shell-regular" >}}

    ```shell
    wget https://raw.githubusercontent.com/pingcap/tidb-operator/v1.2.0-beta.2/manifests/crd.yaml
    ```

    2. Download the `tidb-operator` chart package file.

    {{< copyable "shell-regular" >}}

    ```shell
    wget http://charts.pingcap.org/tidb-operator-v1.2.0-beta.2.tgz
    ```

>     3. Download the Docker image required for the new TiDB Operator version:

Reviewer: ```suggestion
    3. Download the Docker images required for the new TiDB Operator version:
```

    {{< copyable "shell-regular" >}}
    
    ```shell
    docker pull pingcap/tidb-operator:v1.2.0-beta.2
    docker pull pingcap/tidb-backup-manager:v1.2.0-beta.2

>     docker save -o tidb-operator-v1.2.0-beta.2.tar pingcap/tidb-operator:v1.2.0-beta.2

Reviewer: ```suggestion

    docker save -o tidb-operator-v1.2.0-beta.2.tar pingcap/tidb-operator:v1.2.0-beta.2
```

    docker save -o tidb-backup-manager-v1.2.0-beta.2.tar pingcap/tidb-backup-manager:v1.2.0-beta.2
    ```

2. Upload the downloaded files and images to the server that needs to be upgraded, and then take the following steps for installation:

    1. Install the `crd.yaml` file for TiDB Operator:

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f . /crd.yaml
    ```

    2. Unpack the `tidb-operator` chart package file, and then copy the `values.yaml` file:

    {{< copyable "shell-regular" >}}

    ```shell
    tar zxvf tidb-operator-v1.2.0-beta.2.tgz && \
    mkdir -p ${HOME}/tidb-operator/v1.2.0-beta.2 &&
    cp tidb-operator/values.yaml ${HOME}/tidb-operator/v1.2.0-beta.2/values-tidb-operator.yaml
    ```

>     3. Install the Docker image on the server:

Reviewer: ```suggestion
    3. Install the Docker images on the server:
```

    If TiDB Operator is successfully upgraded, the expected output is as follows. `v1.2.0-beta.2` represents the desired version of TiDB Operator.

    ```
    image: pingcap/tidb-operator:v1.2.0-beta.2
    image: docker.io/pingcap/tidb-operator:v1.2.0-beta.2
    image: pingcap/tidb-operator:v1.2.0-beta.2
    image: docker.io/pingcap/tidb-operator:v1.2.0-beta.2
    ```

    > **Note:**
    >
    > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will be automatically upgraded to the corresponding version of TiDB Operator.

## Upgrade TiDB Operator offline

If your server cannot access the Internet, you can take the following steps to upgrade TiDB Operator offline:

1. Download the files and images required for the upgrade using a machine with the Internet access:

    1. Download the `crd.yaml` file for the new TiDB Operator version. For more information about CRD, see [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/).

    {{< copyable "shell-regular" >}}

    ```shell
    wget https://raw.githubusercontent.com/pingcap/tidb-operator/v1.2.0-beta.2/manifests/crd.yaml

>     ```

Reviewer 2: 8 indentations are needed.

>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     wget https://raw.githubusercontent.com/pingcap/tidb-operator/v1.2.0-beta.2/manifests/crd.yaml
>     ```

Reviewer: ```suggestion
        {{< copyable "shell-regular" >}}

        ```shell
        wget https://raw.githubusercontent.com/pingcap/tidb-operator/v1.2.0-beta.2/manifests/crd.yaml
        ```
```

    2. Download the `tidb-operator` chart package file.

>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     wget http://charts.pingcap.org/tidb-operator-v1.2.0-beta.2.tgz
>     ```

Reviewer: ```suggestion
        {{< copyable "shell-regular" >}}

        ```shell
        wget http://charts.pingcap.org/tidb-operator-v1.2.0-beta.2.tgz
        ```
```

    3. Download the Docker images required for the new TiDB Operator version:

>     {{< copyable "shell-regular" >}}
>     
>     ```shell
>     docker pull pingcap/tidb-operator:v1.2.0-beta.2
>     docker pull pingcap/tidb-backup-manager:v1.2.0-beta.2
> 
>     docker save -o tidb-operator-v1.2.0-beta.2.tar pingcap/tidb-operator:v1.2.0-beta.2
>     docker save -o tidb-backup-manager-v1.2.0-beta.2.tar pingcap/tidb-backup-manager:v1.2.0-beta.2
>     ```

Reviewer 2: ```suggestion
8 indentations are needed.
```

>     {{< copyable "shell-regular" >}}
>     
>     ```shell
>     docker pull pingcap/tidb-operator:v1.2.0-beta.2
>     docker pull pingcap/tidb-backup-manager:v1.2.0-beta.2
> 
>     docker save -o tidb-operator-v1.2.0-beta.2.tar pingcap/tidb-operator:v1.2.0-beta.2
>     docker save -o tidb-backup-manager-v1.2.0-beta.2.tar pingcap/tidb-backup-manager:v1.2.0-beta.2
>     ```

Reviewer: ```suggestion
        {{< copyable "shell-regular" >}}
    
        ```shell
        docker pull pingcap/tidb-operator:v1.2.0-beta.2
        docker pull pingcap/tidb-backup-manager:v1.2.0-beta.2

        docker save -o tidb-operator-v1.2.0-beta.2.tar pingcap/tidb-operator:v1.2.0-beta.2
        docker save -o tidb-backup-manager-v1.2.0-beta.2.tar pingcap/tidb-backup-manager:v1.2.0-beta.2
        ```
```

2. Upload the downloaded files and images to the server that needs to be upgraded, and then take the following steps for installation:

    1. Install the `crd.yaml` file for TiDB Operator:

>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     kubectl apply -f . /crd.yaml
>     ```

Reviewer: ```suggestion
        {{< copyable "shell-regular" >}}

        ```shell
        kubectl apply -f . /crd.yaml
        ```
```

    2. Unpack the `tidb-operator` chart package file, and then copy the `values.yaml` file:

>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     tar zxvf tidb-operator-v1.2.0-beta.2.tgz && \
>     mkdir -p ${HOME}/tidb-operator/v1.2.0-beta.2 &&
>     cp tidb-operator/values.yaml ${HOME}/tidb-operator/v1.2.0-beta.2/values-tidb-operator.yaml
>     ```
> 
>     3. Install the Docker images on the server:
> 
>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     docker load -i tidb-operator-v1.2.0-beta.2.tar
>     docker load -i tidb-backup-manager-v1.2.0-beta.2.tar
>     ```

Reviewer: ```suggestion
        {{< copyable "shell-regular" >}}

        ```shell
        tar zxvf tidb-operator-v1.2.0-beta.2.tgz && \
        mkdir -p ${HOME}/tidb-operator/v1.2.0-beta.2 &&
        cp tidb-operator/values.yaml ${HOME}/tidb-operator/v1.2.0-beta.2/values-tidb-operator.yaml
        ```

    3. Install the Docker images on the server:

        {{< copyable "shell-regular" >}}

        ```shell
        docker load -i tidb-operator-v1.2.0-beta.2.tar
        docker load -i tidb-backup-manager-v1.2.0-beta.2.tar
        ```
```

1. Update CustomResourceDefinition (CRD) for Kubernetes. For more information about CRD, see [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/).

    1. Download the `crd.yaml` file using a machine with the Internet access:

    {{< copyable "shell-regular" >}}

     ```shell
     wget https://raw.githubusercontent.com/pingcap/tidb-operator/${version}/manifests/crd.yaml
     ```
    > **Note:**
    >
    > The `${version}` in this document represents the version of TiDB Operator, such as `v1.2.0-beta.2`. You can check the currently supported versions using the `helm search repo -l tidb-operator` command.
    > If the command output does not include the latest version, update the repo using the `helm repo update` command. For details, refer to [Configure the Help repo](tidb-toolkit.md#configure-the-helm-repo).

    2. Upload `crd.yaml` to the server to be upgraded, and then run the following command for the installation and check:

    {{< copyable "shell-regular" >}}

     ```shell
     kubectl apply -f . /crd.yaml && \
     kubectl get crd tidbclusters.pingcap.com
     ```

> 2. Get the `values.yaml` file in the `tidb-operator` chart for the new version of TiDB Operator.
> 
>     1. Download the `tidb-operator` chart package file using a machine with the Internet access:
>     
>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     wget http://charts.pingcap.org/tidb-operator-${version}.tgz
>     ```
> 
>     2. Upload the `tidb-operator` chart package file to the server to be upgraded, and then run the following command:
> 
>     {{< copyable "shell-regular" >}}
> 
>     ```shell
>     tar zxvf tidb-operator-${version}.tgz && \
>     mkdir -p ${HOME}/tidb-operator/${version} && \
>     cp tidb-operator/values.yaml ${HOME}/tidb-operator/${version}/values-tidb-operator.yaml
>     ```

Reviewer: ```suggestion
2. Upload the downloaded files and images to the server that needs to be upgraded, and then take the following steps for installation:

    1. Install the `crd.yaml` file for TiDB Operator:

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f . /crd.yaml
    ```

    2. Unpack the `tidb-operator` chart package file, and then copy the `values.yaml` file:

    {{< copyable "shell-regular" >}}

    ```shell
    tar zxvf tidb-operator-v1.2.0-beta.2.tgz && \
    mkdir -p ${HOME}/tidb-operator/v1.2.0-beta.2 &&
    cp tidb-operator/values.yaml ${HOME}/tidb-operator/v1.2.0-beta.2/values-tidb-operator.yaml
    ```

    3. Install the Docker image on the server:

    {{< copyable "shell-regular" >}}

    ```shell
    docker load -i tidb-operator-v1.2.0-beta.2.tar
    docker load -i tidb-backup-manager-v1.2.0-beta.2.tar
    ```
```

> 3. Download the following Docker images for the upgrade using a machine with the Internet access, upload the downloaded images to the server to be upgraded, and then install the following Docker images using the `docker load` command. For detailed commands of downloading and installing Docker images, see [Download the Docker images used by TiDB Operator](deploy-tidb-operator.md#download-the-docker-images-used-by-tidb-operator).
> 
>    ```shell
>     pingcap/tidb-operator:${version}
>     pingcap/tidb-backup-manager:${version}
>    ```
> 

Reviewer: ```suggestion
```

> 4. In the `${HOME}/tidb-operator/${version}/values-tidb-operator.yaml` file, modify the `operatorImage` version to the new TiDB Operator version. Merge the customized configuration in the old `values.yaml` file to the `${HOME}/tidb-operator/${version}/values-tidb-operator.yaml` file, and then execute `helm upgrade`:

Reviewer: ```suggestion
3. In the `${HOME}/tidb-operator/${version}/values-tidb-operator.yaml` file, modify the `operatorImage` version to the new TiDB Operator version. Merge the customized configuration in the old `values.yaml` file to the `${HOME}/tidb-operator/${version}/values-tidb-operator.yaml` file, and then execute `helm upgrade`:
```

   {{< copyable "shell-regular" >}}

    ```shell

>     helm upgrade tidb-operator ./tidb-operator --version=${version} -f ${HOME}/tidb-operator/${version}/values-tidb-operator.yaml

Reviewer: ```suggestion
    helm upgrade tidb-operator ./tidb-operator --version=v1.2.0-beta.2 -f ${HOME}/tidb-operator/v1.2.0-beta.2/values-tidb-operator.yaml
```

    >
    > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will automatically upgrade to the specified version of TiDB Operator.

> ## Offline Upgrade TiDB Operator

Reviewer: PTAL. Thanks.

    ```

3. Download `operatorImage` image version on an Internet machine and upload to server, then use `docker load` to install Docker image to server (for details, please refer to [Download the Docker images used by TiDB Operator](deploy-tidb-operator.md)), then upgrade via the following command:

   {{< copyable "shell-regular" >}}

    ```shell
    helm upgrade tidb-operator ./tidb-operator --version=${version} -f ${HOME}/tidb-operator/${version}/values-tidb-operator.yaml
    ```

   After all the Pods start normally, execute the following command to check the image of TiDB Operator:

   {{< copyable "shell-regular" >}}

    ```shell
    kubectl get po -n tidb-admin -l app.kubernetes.io/instance=tidb-operator -o yaml | grep 'image:.*operator:'
    ```

   If TiDB Operator is successfully upgraded, the expected output is as follows. `${version}` represents the desired version of TiDB Operator.

    ```
    image: pingcap/tidb-operator:${version}
    image: docker.io/pingcap/tidb-operator:${version}
    image: pingcap/tidb-operator:${version}
    image: docker.io/pingcap/tidb-operator:${version}
    ```

   > **Note:**
   >

>    > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will automatically upgrade to the specified version of TiDB Operator.

Reviewer: ```suggestion
   > After TiDB Operator is upgraded, the `discovery` deployment in all TiDB clusters will be automatically upgraded to the specified version of TiDB Operator.
```
